### PR TITLE
added feature toggle mode

### PR DESCRIPTION
**Pull request by comzine**

### CODE REVIEW

File: uhubctl.c [hunk 1]
```diff
@@ -1051,9 +1058,12 @@ int main(int argc, char *argv[])
                 for (port=1; port <= hubs[i].nports; port++) {
                     if ((1 << (port-1)) & ports) {
                         int port_status = get_port_status(devh, port);
+                        printf("Port Status: %i\n", port_status);
```
Comment (mvp) [inline uhubctl.c:1061]: We don't want to always print port status here. Is this left from your debugging?

File: uhubctl.c [hunk 2]
```diff
@@ -1051,9 +1058,12 @@ int main(int argc, char *argv[])
...
+                        if (opt_action == POWER_TOGGLE)
```
Comment (mvp) [inline uhubctl.c:1062]: Please use extra braces for `if`

File: uhubctl.c [hunk 3]
```diff
@@ -1051,9 +1058,12 @@ int main(int argc, char *argv[])
...
+                            request = (port_status == 0) ? LIBUSB_REQUEST_SET_FEATURE : LIBUSB_REQUEST_CLEAR_FEATURE;
```
Comment (mvp) [inline uhubctl.c:1063]: `port_status == 0` does not necessarily mean power is off.
You need to use `port_status & power_mask` instead

File: uhubctl.c [hunk 4]
```diff
@@ -1051,9 +1058,12 @@ int main(int argc, char *argv[])
...
                         int power_mask = hubs[i].super_speed ? USB_SS_PORT_STAT_POWER
                                                              : USB_PORT_STAT_POWER;
-                        if (k == 0 && !(port_status & power_mask))
+                        if (k == 0 && !(port_status & power_mask) && (opt_action != POWER_TOGGLE))
```
Comment (mvp) [inline uhubctl.c:1066]: This condition does not seem right. I assume you meant this?

    if (k == 0 && (!(port_status & power_mask) || opt_action != POWER_TOGGLE))

File: uhubctl.c [hunk 5]
```diff
@@ -1029,6 +1033,9 @@ int main(int argc, char *argv[])
             continue;
         if (k == 1 && opt_action == POWER_KEEP)
             continue;
+        // if power_toggle is active: only one time
```
Comment (mvp) [inline uhubctl.c:1036]: Change comment to:

    // if toggle requested, do it only once when `k == 0`

File: README.md [hunk 6]
```diff
@@ -165,7 +165,7 @@ You can control the power on a USB port(s) like this:
     uhubctl -a off -p 2
 
 This means operate on default smart hub and turn power off (`-a off`, or `-a 0`)
-on port 2 (`-p 2`). Supported actions are `off`/`on`/`cycle` (or `0`/`1`/`2`).
+on port 2 (`-p 2`). Supported actions are `off`/`on`/`cycle`/`toggle` (or `0`/`1`/`2`/`3`).
```
Comment (mvp) [inline README.md:168]: Please add toggle mention to `print_usage()` in `uhubctl.c`

File: uhubctl.c (hunk 4 above)
Comment (comzine) [reply uhubctl.c:1066]: I don't think that this one is wrong, as it should only "continue", if power_toggle is not requested AND the other conditions are met.

File: uhubctl.c (hunk 1 above)
Comment (comzine) [reply uhubctl.c:1061]: Of course - yes.

File: uhubctl.c (hunk 2 above)
Comment (comzine) [reply uhubctl.c:1062]: Ok

File: uhubctl.c (hunk 3 above)
Comment (comzine) [reply uhubctl.c:1063]: Ok, just tested it with port_status and it worked. I'll change …

File: uhubctl.c (hunk 5 above)
Comment (comzine) [reply uhubctl.c:1036]: Ok

File: README.md (hunk 6 above)
Comment (comzine) [reply README.md:168]: Of course, yes.

File: uhubctl.c [hunk 7]
```diff
@@ -1053,7 +1060,10 @@ int main(int argc, char *argv[])
                         int port_status = get_port_status(devh, port);
                         int power_mask = hubs[i].super_speed ? USB_SS_PORT_STAT_POWER
                                                              : USB_PORT_STAT_POWER;
-                        if (k == 0 && !(port_status & power_mask))
+                        if (opt_action == POWER_TOGGLE) {
+                            request = (port_status & power_mask) ? LIBUSB_REQUEST_SET_FEATURE : LIBUSB_REQUEST_CLEAR_FEATURE;
```
Comment (mvp) [inline uhubctl.c:1064]: I tested this locally, this condition is opposite to what it should be. Please change to following (works for me):
```
request = (port_status & power_mask) ? LIBUSB_REQUEST_CLEAR_FEATURE
                                     : LIBUSB_REQUEST_SET_FEATURE;
```

**Review by mvp — CHANGES_REQUESTED** [review]
Thank you @comzine!
This is all looking excellent, with just one issue to address (opposite condition for request when toggling).

File: uhubctl.c (hunk 7 above)
Comment (comzine) [reply uhubctl.c]: Thank you 👍! Resolved …

**Comment by mvp** [comment]
Thank you for your contribution! Merged

**Pull request merged**